### ISSUE TITLE
Allowing $ref in spec

### ISSUE THREAD

**Issue by emilianozublena** [hotmeteor/spectator#84]
Hello everyone! first of all, great package! not sure if this is allowed by the library (and the underlying parser) but, will it read through $ref's as long as they're valid within open api specs?

Thanks everyone!

**Comment by jarrodparkes** [comment]
@emilianozublena thank you for the kind words.

Yes, this library should read `$ref`s. However, if you have a circular reference, then it stops short of validating the circular reference so that it doesn't hit an infinite loop.

Have you been experiencing otherwise?

**Comment by emilianozublena** [comment]
Not so far, we made it work, though it's really hard to identify if "something" is wrong w the spec... we're having issues were the library is exploding but no meaningful errors are returned (and the validators we're using against the spec are green flagging it and all the generators from both swagger & openapi community driven ones are working).

Any ideas? can share an example of the spec, though its a bit big to just paste it in here.

Thanks again!!

**Comment by jarrodparkes** [comment]
@emilianozublena what version are you using? we recently made a bunch of changes in v1.4.0 that should print meaningful errors. can you share the current error output? I'd expect to see the expected schema with the errors alongside

**Comment by mjt2021** [comment]
@jarrodparkes hey there, we are indeed using v1.4

let me give you some examples of things we've been facing:
- if there are relational paths written like this "./../" it just fails with an "undefined offset"
- this discussion: https://github.com/hotmeteor/spectator/discussions/86

if fixed a small issue in SchemaValidationException.php : 
 // is this a polymorphic schema?
        $polymorphic_keys = array_filter($keys_at_location, function ($key) {
            return $key == 'allOf' || $key == 'anyOf' || $key == 'oneOf';
        }, ARRAY_FILTER_USE_KEY);
        $polymorphic_keys = array_flip($polymorphic_keys);
        $polymorphic_keys = array_values($polymorphic_keys);    <<<< added this line cause the array had 1 item with key name 1

        if (! empty($polymorphic_keys)) { // first, check for a polymorphic schema...
            $polymorphic_key = $polymorphic_keys[0];   <<<< failed here with undefined offset 0, cause it was just [1=>allOf]

**Comment by jarrodparkes** [comment]
@mjt2021 hmm, that wasn't quite what I was expecting. Have you been able to produce any failure output that looks like this?

```
  ---

The properties must match schema: data
All array items must match schema
The required properties (name) are missing

object++ <== The properties must match schema: data
    status*: string
    data*: array <== All array items must match schema
        object <== The required properties (name) are missing
            id*: string
            name*: string
            slug: string?

  ---
```

Or is it the case that we are failing before any validation actually occurs?

**Comment by mjt2021** [comment]
still no luck reaching that level of detailed error. i do get "url not found in spec" but if found, i'm facing some issues

**Comment by jarrodparkes** [comment]
@mjt2021 is this the same thing referred to in #86 ? Did you solve it?

**Comment by mjt2021** [comment]
it is not the same. important thing here i guess is this piece of code
          
                    $polymorphic_keys = array_flip($polymorphic_keys);
                    $polymorphic_keys = array_values($polymorphic_keys);    <<<< added this line cause the array had 1 item with key name 1
          
            if (! empty($polymorphic_keys)) { // first, check for a polymorphic schema...
                $polymorphic_key = $polymorphic_keys[0];   <<<< failed here with undefined offset 0, cause it was just [1=>allOf]

**Comment by jarrodparkes** [comment]
@mjt2021 and what did changing this line of code do? can you provide a full example (files) so I could reproduce? it sounds like you are suggesting a fix, but I'm still not really sure what the fix is doing

**Comment by mjt2021** [comment]
Sorry that I can't provide enough information on the spec, but I can share the piece that is giving me issues

This spec:
        ResponseWithPagination:
              title: Base response w pagination
              allOf:
                - $ref: '#/components/schemas/BaseResponse'
                - type: object
                - required:
                  - paging
                  properties:
                    paging:
                      $ref: '#/components/schemas/PaginationResponse'

When goes through this piece of code
         $keys_at_location = array_flip(array_keys($schema));

Produces this array:
        array:2 [
          "title" => 0
          "allOf" => 1
        ]

And so this line:

        $polymorphic_key = $polymorphic_keys[0];

Fails, because after the next filter + flip, there will be no 0 index, but a 1 index

I just added the array_values because it resets the array indexes, it will always be 0,1,2 etc

I hope this helps

**Comment by jarrodparkes** [comment]
@mjt2021 This is still not quite everything I need to reproduce the issue. Can you share an example of the test/asserting you are running. If security is an issue, just omit anything you don't want to share publicly. Also, can you format the YML (using back ticks)? That would make this a lot easier to grok.

```yml
title: Example
```

**Comment by mjt2021** [comment]
This error happened during the `assertValidResponse(200)`

Actual test is very simple. The spec is very large and has lots of references
```
Spectator::using('SPEC.yaml');
$this->app[Kernel::class]->prependMiddlewareToGroup('api_v1', Middleware::class);  <<<<< this one is important, we don't use the default middleware group, so the lib was not even doing a thing. maybe add in the setup doc that if you use a different middleware group you got to hook it (maybe you have a more elegant way :)

$this->getJson('/api/v1/vendors?customerLocationId=' . $customerLocation->id);

return $response->assertValidRequest()
                        ->assertValidResponse(200);
```
Let me know if this is enough. Otherwise, I can try to do an atomic example of the issue, and I'll share the files then. Sorry again for the trouble :)

Hey have a nice holiday today! If you celebrate anything at all!

**Comment by jarrodparkes** [comment]
thanks @mjt2021 I'll be back in the office Monday and can take a look then

**Comment by hotmeteor** [comment]
@jarrodparkes Do you need any help debugging this?

**Comment by jarrodparkes** [comment]
@hotmeteor yes please 🙏

**Comment by ArondeParon** [comment]
> ........
> And so this line:
> 
> ```
>               $polymorphic_key = $polymorphic_keys[0];
> ```
> 

Experiencing the same issue. The exception doesn't provide any more details so it's a bit hard to debug without doing a deep dive into the the internals, but as it seems right now, it would make sense to wrap the function in `array_values` so that the `0` key is always present.

**Comment by hotmeteor** [comment]
Feel free to create a PR!

**Comment by beblife** [comment]
@hotmeteor I think my PR here #101 would fix this issue!

**Comment by shortcodes** [comment]
@beblife is your PR resolves the issue? I got newest version and testing still passes without errors if $ref is part of schema but it shoud not.

**Comment by beblife** [comment]
> @beblife is your PR resolves the issue? I got newest version and testing still passes without errors if $ref is part of schema but it shoud not.

The PR only fixes outputting the exception message when dealing with nested reference paths, e.g `../../some-file.json`. There is nothing in there regarding validation of a schema and is most likely not an issue with this package but maybe of the `cebe/php-openapi` dependency that does the validation.

**Comment by beblife** [comment]
@hotmeteor @jarrodparkes Wondering if this issue can be closed? Given that PR #101 is merged.
